deepchecks/nlp/checks/data_integrity/frequent_substrings.py

```diff
@@ -0,0 +1,417 @@
+        """
+        diff = df['Frequency'].diff().abs()
+        ratio = diff.div(df['Frequency'].shift(1))
+        ratio[ratio <= self.min_relative_change] = 0
+        max_peak = ratio.max()
+        if max_peak == 0:
+            return len(df)
+        return list(ratio[ratio == max_peak].index)[0]
+
+    def _filter_peak_and_significant(self, df):
+        """
+        Filter the dataframe to retain only the significant substrings.
+
+        The function first identifies the cut-off index based on the significance ratio. Then, it identifies
+        a cut-off based on the peak in the frequency ratio. The final dataframe is truncated using the maximum
+        of these two cut-off indices.
+
+        Parameters
+        ----------
+        df : DataFrame
+            The input dataframe expected to have a 'Frequency' column.
+
+        Returns
+        -------
+        DataFrame
+            A truncated dataframe containing only the significant substrings based on the established criteria.
+        """
+        if len(df) == 1:
+            return df
+        significant_cut_ind = self._get_significant_cut_ind(df)
+        peak_cut_ind = self._identify_peak_cut(df)
+        return df[:max(significant_cut_ind, peak_cut_ind)]
+
+    def run_logic(self, context: Context, dataset_kind):
+        """Run check.
+
+        Parameters:
+        -----------
+        context : Context
+            Contains dataset and related methods.
+        dataset_type :
+            Type or format of the dataset.
+
+        Returns
+        -------
+        CheckResult
+            Results containing frequent substrings' information.
+
+        Raises
+        ------
+        DeepchecksValueError
+            If the Dataset is empty.
+        """
+        dataset = context.get_data_by_kind(dataset_kind)
+        num_samples = len(dataset)
+        dataset = dataset.sample(self.n_samples, random_state=self.random_state)
+        dataset = t.cast(TextData, dataset)
+        if dataset.n_samples == 0:
+            raise DeepchecksValueError('Dataset cannot be empty')
+        data = list(zip(dataset.get_original_text_indexes(), dataset.text))
+
+        substrings_dict = self._find_frequent_substrings(data)
+        substrings_dict = self._eliminate_overlapping_substrings(substrings_dict)
+
+        if len(substrings_dict) == 0:
+            value = {}
+            display = None
+
+        else:
+            sorted_substrings = sorted(substrings_dict.items(), key=lambda x: (x[1]['freq'], x[0]), reverse=True)
+            df = pd.DataFrame({
+                'Text': [item[0] for item in sorted_substrings],
+                'Frequency': [item[1]['freq'] for item in sorted_substrings],
+                'Sample IDs': [item[1]['original_indexes'] for item in sorted_substrings]
+            })
+
+            df = self._filter_peak_and_significant(df)
+
+            if self.n_samples < num_samples:
+                dataset = context.get_data_by_kind(dataset_kind)
```

Comment:
if we didn't override the dataset with the sampled one we could have avoided this